### PR TITLE
Add validation to disk's name in CreateVMWizard Storage step

### PR DESCRIPTION
**Pull request by bamsalem**
This PR adds validations to disk name such as in NIC name in Network Step.
The rules for the validation taken from [the documentation](https://access.redhat.com/documentation/en-us/red_hat_virtualization/4.4/html/administration_guide/sect-virtual_disk_tasks).
In case of error it will show error message, mark the field in red and not let user save the edited disk and move to the next step:

image

If you have any questions please feel free to ask
Fixes: #1329

### CODE REVIEW

**Comment by ovirt-infra** [comment]
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically. ( keep in mind this list might be overwritten if the job XML is refreshed, for permanent whitelisting, please follow #3 option )
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

**Comment by sgratch** [comment]
ci add to whitelist

File: src/intl/messages.js [hunk 1]
```diff
@@ -167,6 +167,7 @@ export const messages: { [messageId: string]: MessageType } = {
   createVmWizardStepTitleNetwork: 'Networking',
   createVmWizardStepTitleReview: 'Review',
   createVmWizardStepTitleStorage: 'Storage',
+  createVmWizardStorageDiskNameRules: 'Disk name must be 40 or less alphanumeric characters or "-_."',
```
Comment (sgratch) [inline src/intl/messages.js:170]: This message is not accurate, please check that you can create a disk name longer than 40 chars. Please rephrase it to be the same as on webadmin:
Disk name can contain only alphanumeric, '.', '_' or '-' characters.

File: src/components/utils/validation.js [hunk 2]
```diff
@@ -88,7 +88,7 @@ export function isVmNameValid (nameCandidate: string): boolean {
 }
 
 export function isDiskNameValid (nameCandidate: string): boolean {
-  return isVmNameValid(nameCandidate)
+  return /^[a-zA-Z0-9_\-\\.]{1,40}$/.test(nameCandidate)
```
Comment (sgratch) [inline src/components/utils/validation.js:91]: Please update  this length validation "{1,40}" to a much longer length (e.g. 64 chars)

**Review by sgratch — DISMISSED** [review]
Please see inline comments.

In addition, there is a problem of saving a new disk after an error is displayed:
1. create  a new disk with invalid name - validation failed and an inline error is displayed.
2. fix the disk name to become a valid one.

Expected: saving the disk is possible.
Actual behaviour: no error is displayed but you can't save the disk

File: src/intl/messages.js (hunk 1 above)
Comment (bamsalem) [reply src/intl/messages.js:170]: image

File: src/components/utils/validation.js (hunk 2 above)
Comment (bamsalem) [reply src/components/utils/validation.js:91]: @sgratch Are you sure?
I stuck to the [documentation](https://access.redhat.com/documentation/en-us/red_hat_virtualization/4.3/html/administration_guide/sect-virtual_disk_tasks#Creating_a_Virtual_Disk)

Comment (sgratch) [reply src/components/utils/validation.js:91]: > @sgratch Are you sure?
> I stuck to the [documentation](https://access.redhat.com/documentation/en-us/red_hat_virtualization/4.3/html/administration_guide/sect-virtual_disk_tasks#Creating_a_Virtual_Disk)

When comparing webadmin code/behaviour to the documentation, usually the webadmin behaviour is the the enforcing rule (and probably the doc is not up to date) :). 
Specifically here, the limit is 255 chars:
https://github.com/oVirt/ovirt-engine/blob/e48484ccb3c1c8fcaed81023b234295a109ebd8b/frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/DisksAllocationModel.java#L486-L490

Comment (bamsalem) [reply src/components/utils/validation.js:91]: OK, I'll check webAdmin code next time :)

File: src/components/utils/validation.js [hunk 3]
```diff
@@ -88,7 +88,7 @@ export function isVmNameValid (nameCandidate: string): boolean {
 }
 
 export function isDiskNameValid (nameCandidate: string): boolean {
-  return isVmNameValid(nameCandidate)
+  return /^[a-zA-Z0-9_\-\\.]{1,64}$/.test(nameCandidate)
```
Comment (sgratch) [inline src/components/utils/validation.js:91]: Why limiting to 64 chars while on webadmin and backend we now know that the limitation is for 255 chars? It's better to use the same limitation

**Review by sgratch — CHANGES_REQUESTED** [review]

File: src/components/CreateVmWizard/steps/Storage.js [hunk 4]
```diff
@@ -503,21 +517,32 @@ class Storage extends React.Component {
...
+        if (value >= minDiskSizeInGiB && value <= maxDiskSizeInGiB) {
+          delete editingRow.invalidSizeValue
+        } else {
+          editingRow.invalidSizeValue = true
+        }
+        value = +value * (1024 ** 3) // GiB to B
+        break
+      case 'name':
+        editingErrors.diskInvalidName = !isDiskNameValid(value)
+        console.log('editingErrors', editingErrors)
```
Comment (sgratch) [inline src/components/CreateVmWizard/steps/Storage.js:535]: Don't you need to delete this console.log line? I can't see the benefit of leaving it.

**Review by sgratch — COMMENTED** [review]
A small inline comment.

File: src/components/utils/validation.js [hunk 5]
```diff
@@ -88,7 +88,7 @@ export function isVmNameValid (nameCandidate: string): boolean {
 }
 
 export function isDiskNameValid (nameCandidate: string): boolean {
-  return isVmNameValid(nameCandidate)
+  return /^[a-zA-Z0-9_\-\\.]{1,255}$/.test(nameCandidate)
```
Comment (rszwajko) [inline src/components/utils/validation.js:91]: Please add an unit test  - previously `isVmNameValid()` was unit tested.
@sgratch  @sjd78 
Validation is now more restrictive. However `isDiskNameValid` is also used to validate alias name in `DiskImageEditor`. Can we break some use case here?

**Review by rszwajko — DISMISSED** [review]

File: src/components/CreateVmWizard/steps/Storage.js [hunk 6]
```diff
@@ -503,21 +517,32 @@ class Storage extends React.Component {
...
-      }
-      value = +value * (1024 ** 3) // GiB to B
+    const editingErrors = {}
+    switch (field) {
+      case 'size':
+        if (!isNumber(value)) return
+
+        const { minDiskSizeInGiB, maxDiskSizeInGiB } = this.props
+        if (value >= minDiskSizeInGiB && value <= maxDiskSizeInGiB) {
+          delete editingRow.invalidSizeValue
```
[the remark this answers is not in the document]
Comment (bamsalem) [reply src/components/CreateVmWizard/steps/Storage.js:527]: it was part of another fix, I'll change that

File: src/components/CreateVmWizard/steps/Storage.js [hunk 7]
```diff
@@ -105,6 +107,9 @@ class Storage extends React.Component {
     this.isEditingMode = this.isEditingMode.bind(this)
 
     this.state = {
+      editingErrors: {
```
[the remark this answers is not in the document]
Comment (bamsalem) [reply src/components/CreateVmWizard/steps/Storage.js:110]: `editing` is an object that includes the data of the edited/created disk, and `creating`  just holds the id of the created disk in case of editing its value is false, so the answer is no, hope I explained it well :)

Creating disk:
creating
Editing disk:
editing

File: src/components/utils/validation.js (hunk 5 above)
Comment (bamsalem) [reply src/components/utils/validation.js:91]: I added unit test :)

Comment (sgratch) [reply src/components/utils/validation.js:91]: @bamsalem @rszwajko regarding using isDiskNameValid in both create/edit modes:
we can't break any use case since isVmNameValid and isDiskNameValid actually share the same validation logic except length limitation, so now validating of disk alias name in DiskImageEditor is correct.
But two issues just bumped up:
-  is there any reason for using VM_NAME_REGEX.test(nameCandidate) within isVmNameValid function and  /^[a-zA-Z0-9_\-\\.]{1,255}$/.test(nameCandidate) within isDiskNameValid function? We can check the length limitation separately and share the same regex...

- for both DiskImageEditor and Storage.js we use the same validation but display a different error message (pleaseEnterValidDiskName and createVmWizardStorageDiskNameRules). I suggest using the same message.

**Review by sgratch — DISMISSED** [review]

File: src/components/utils/validation.js (hunk 5 above)
Comment (rszwajko) [reply src/components/utils/validation.js:91]: @sgratch  my understanding is that VM_NAME_REGEXP is **less** strict - check [UNICODE_NON_ASCII_LETTERS](https://github.com/oVirt/ovirt-web-ui/blob/a41b92f1e340480129c2cd5f6c9da58066ef96ce/src/components/utils/validation.js#L64)

Comment (bamsalem) [reply src/components/utils/validation.js]: @sgratch
I fixed the validations now we are using only the /^[a-zA-Z0-9_-\.]{1,255}$/.test(nameCandidate) validation.

Regarding the message we display different messages because the message in Create VM wizard must be shorter than DiskImageEditor, that more descriptive:

Long message:
image

Short message:
image

If you still think we should use the same message, WDYT? which one is preferred?

File: src/components/utils/validation.test.js [hunk 8]
```diff
@@ -20,6 +20,28 @@ describe('check host names', function () {
   })
 })
 
+describe('check disk names', function () {
+  it('valid disk names', function () {
+    expect(isDiskNameValid('veryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryValidDiskName12')).toEqual(true)
```
Comment (sjd78) [inline src/components/utils/validation.test.js:25]:
```suggestion
    expect(isDiskNameValid('abcde'.repeat(51))).toEqual(true) // 255 characters
```

**Review by sjd78 — DISMISSED** [review]

**Comment by sjd78** [comment]
@sgratch , @bamsalem , @rszwajko - The VM name regexp needs to be changed back.  VMs can have names that do not contain latin alpha characters.  In fact, the regexp is a copy from webadmin's VM name validation.

I'm not sure if disk names can contain non-latin alpha characters.  I did not test that, or lookup the validation code, yet.

File: src/components/utils/validation.test.js [hunk 9]
```diff
@@ -18,8 +18,30 @@ describe('check host names', function () {
...
+    expect(isDiskNameValid('-abc')).toEqual(true)
+  })
+
+  it('invalid disk names', function () {
+    expect(isDiskNameValid('')).toEqual(false)
+    expect(isDiskNameValid('a b')).toEqual(false)
+    expect(isDiskNameValid('a!@$b298')).toEqual(false)
+    expect(isDiskNameValid('a_name()')).toEqual(false)
+    expect(isDiskNameValid('Кирилиця')).toEqual(false)
+    expect(isDiskNameValid('veryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryveryverylongDiskName')).toEqual(false)
```
[the remark this answers is not in the document]
Comment (rszwajko) [reply src/components/utils/validation.test.js:41]: How about this: `'a'.repeat(256)`? we want to check exactly one above the limit

[the remark this answers is not in the document]
Comment (sjd78) [reply src/components/utils/validation.test.js:41]: > How about this: `'a'.repeat(256)`? we want to check exactly one above the limit

Even better

File: src/components/utils/validation.js [hunk 10]
```diff
@@ -1,97 +1,42 @@
 // @flow
 
+const charactersRegExpTest = (nameCandidate: string) => /^[a-zA-Z0-9_\-\\.]*$/.test(nameCandidate)
```
Comment (rszwajko) [inline src/components/utils/validation.js:3]: @sgratch  Actually I'm not sure if we should throw away UTF chars. Check  I18NNameValidation which is used DiskAllocationModel - it [defines letters as all chars including UTF](https://github.com/oVirt/ovirt-engine/blob/e48484ccb3c1c8fcaed81023b234295a109ebd8b/frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/validation/I18NNameValidation.java#L11)

**Review by rszwajko — CHANGES_REQUESTED** [review]

**Comment by sgratch** [comment]
@bamsalem @sjd78 @rszwajko 

> is there any reason for using VM_NAME_REGEX.test(nameCandidate) within isVmNameValid function and /^[a-zA-Z0-9_-.]{1,255}$/.test(nameCandidate) within isDiskNameValid function? We can check the length limitation separately and share the same regex...

You are all right, based on the fact that disk alias string is auto generated from the vm name ({vmName}_Disk1) then we should support UTF for both disk name and vm name.

And by checking the code: 
disk: https://github.com/oVirt/ovirt-engine/blob/e48484ccb3c1c8fcaed81023b234295a109ebd8b/frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/storage/DisksAllocationModel.java#L486-L490

vm: https://github.com/oVirt/ovirt-engine/blob/9059df4d26f5d327c87495ee9e19cb1876ad92b7/frontend/webadmin/modules/uicommonweb/src/main/java/org/ovirt/engine/ui/uicommonweb/models/vms/VmModelBehaviorBase.java#L1665-L1667

Both use _I18NNameValidation_().

So in our web-ui code we should use VM_NAME_REGEX.test(nameCandidate) for both validations


> Regarding the message we display different messages because the message in Create VM wizard must be shorter than DiskImageEditor, that more descriptive

I'm totally ok with having the shorter message version for both dialogs. They both cover the same info and it's better to maintain one message instead of two (if validation login will be ever changed).

**Comment by bamsalem** [comment]
I changed it to use VM_NAME_REGEX again in diskName and vmName.
I changed the code according to your review @sgratch @sjd78 @rszwajko...


> I'm totally ok with having the shorter message version for both dialogs. They both cover the same info and it's better to maintain one message instead of two (if validation login will be ever changed).

Now the same message displayed in CreateVm and editVm:
image

**Comment by sgratch** [comment]
@bamsalem please rebase

**Review by sjd78 — APPROVED** [review]
Component code and error messages look good.
Validation checks and tests look good as well.

**Comment by bamsalem** [comment]
@sgratch 
rebased.

**Comment by isaranova** [comment]
LGTM
